File: server/hotrod/src/main/scala/org/infinispan/server/hotrod/HotRodUtils.scala
package org.infinispan.server.hotrod

import java.io.{IOException, StreamCorruptedException}
import java.lang.StringBuilder

class UnknownVersionException(reason: String, val version: Byte, val messageId: Long)
extends StreamCorruptedException(reason)

class HotRodUnknownOperationException(reason: String, val version: Byte, val messageId: Long)
extends UnknownOperationException(reason)

class InvalidMagicIdException(reason: String) extends StreamCorruptedException(reason)

class CacheUnavailableException extends Exception

class RequestParsingException(reason: String, val version: Byte, val messageId: Long, cause: Exception)
extends IOException(reason, cause) {
   def this(reason: String, version: Byte, messageId: Long) = this(reason, version, messageId, null)
}

class HotRodHeader {
   var op: HotRodOperation = _
   var version: Byte = _
   var messageId: Long = _
   var cacheName: String = _
   var flag: Int = _
   var clientIntel: Short = _
   var topologyId: Int = _

   override def toString = {
      new StringBuilder().append("HotRodHeader").append("{")
      .append("op=").append(op)
      .append(", version=").append(version)
      .append(", messageId=").append(messageId)
      .append(", cacheName=").append(cacheName)
      .append(", flag=").append(flag)
      .append(", clientIntelligence=").append(clientIntel)
      .append(", topologyId=").append(topologyId)
      .append("}").toString
   }
}

class CacheNotFoundException(msg: String, override val version: Byte, override val messageId: Long)
extends RequestParsingException(msg, version, messageId)

class HotRodException(val response: ErrorResponse, cause: Throwable) extends Exception(cause)

class UnknownOperationException(reason: String) extends StreamCorruptedException(reason)

File: server/hotrod/src/main/java/org/infinispan/server/hotrod/HotRodUtils.java
package org.infinispan.server.hotrod;

import java.io.StreamCorruptedException;

import org.infinispan.commons.CacheException;
import org.infinispan.commons.dataconversion.MediaType;

class UnknownVersionException extends StreamCorruptedException {
   final byte version;
   final long messageId;

   public UnknownVersionException(String cause, byte version, long messageId) {
      super(cause);
      this.version = version;
      this.messageId = messageId;
   }

   public HotRodHeader toHeader() {
      return new HotRodHeader(HotRodOperation.ERROR, version, messageId, "", 0, (short) 1, 0, MediaType.MATCH_ALL, MediaType.MATCH_ALL, null);
   }
}

class HotRodUnknownOperationException extends UnknownOperationException {
   final byte version;
   final long messageId;

   public HotRodUnknownOperationException(String cause, byte version, long messageId) {
      super(cause);
      this.version = version;
      this.messageId = messageId;
   }

   public HotRodHeader toHeader() {
      return new HotRodHeader(HotRodOperation.ERROR, version, messageId, "", 0, (short) 1, 0, MediaType.MATCH_ALL, MediaType.MATCH_ALL, null);
   }
}

class InvalidMagicIdException extends StreamCorruptedException {
   public InvalidMagicIdException(String cause) {
      super(cause);
   }
}

class CacheUnavailableException extends CacheException {

}

class RequestParsingException extends CacheException {
   final byte version;
   final long messageId;

   public RequestParsingException(String reason, byte version, long messageId, Exception cause) {
      super(reason, cause);
      this.version = version;
      this.messageId = messageId;
   }

   public RequestParsingException(String reason, byte version, long messageId) {
      super(reason);
      this.version = version;
      this.messageId = messageId;
   }

   public HotRodHeader toHeader() {
      return new HotRodHeader(HotRodOperation.ERROR, version, messageId, "", 0, (short) 1, 0, MediaType.MATCH_ALL, MediaType.MATCH_ALL, null);
   }
}

class CacheNotFoundException extends RequestParsingException {
   public CacheNotFoundException(String reason, byte version, long messageId) {
      super(reason, version, messageId);
   }
}

class UnknownOperationException extends StreamCorruptedException {
   public UnknownOperationException(String reason) {
      super(reason);
   }
}
